src/value.rs

```diff
@@ -158,21 +159,36 @@ pub enum PrimValKind {
 }
 
 impl<'a, 'tcx: 'a> Value {
-    pub(super) fn read_ptr(&self, mem: &Memory<'a, 'tcx>) -> EvalResult<'tcx, Pointer> {
+    #[inline]
+    pub(super) fn by_ref(ptr: Pointer) -> Self {
+        Value::ByRef(ptr, true)
+    }
+
+    /// Convert the value into a pointer (or a pointer-sized integer).  If the value is a ByRef,
+    /// this may have to perform a load.
+    pub(super) fn into_ptr(&self, mem: &mut Memory<'a, 'tcx>) -> EvalResult<'tcx, Pointer> {
         use self::Value::*;
         match *self {
-            ByRef(ptr) => mem.read_ptr(ptr.to_ptr()?),
+            ByRef(ptr, aligned) => {
+                mem.begin_unaligned_read(aligned);
+                let r = mem.read_ptr(ptr.to_ptr()?);
+                mem.end_unaligned_read();
+                r
+            },
             ByVal(ptr) | ByValPair(ptr, _) => Ok(ptr.into()),
         }
     }
 
-    pub(super) fn expect_ptr_vtable_pair(
+    pub(super) fn into_ptr_vtable_pair(
         &self,
         mem: &Memory<'a, 'tcx>
     ) -> EvalResult<'tcx, (Pointer, MemoryPointer)> {
         use self::Value::*;
         match *self {
-            ByRef(ref_ptr) => {
+            ByRef(ref_ptr, aligned) => {
+                if !aligned {
+                    return Err(EvalError::Unimplemented(format!("Unaligned fat-pointers are not implemented")));
```

Comment:
nit: "are not implemented" is unnecessary, the error already prints something along those lines